### PR TITLE
Display folders filter as a directory tree

### PR DESCRIPTION
**Pull request by Nazar65**
Added jquery.jstrre, and service wich deliver directory listing for
jstree plugin

### Fixed Issues (if relevant)

1. magento/adobe-stock-integration#831: Display folders filter as a directory tree
2. ...

### Preconditions
1. Install Magento with Adobe Stock Integration

### Steps to reproduce (*)
1. Login to Admin Panel
2. Select "Media Gallery Development" from the menu on the right
3. Expand filters section

### Actual result (*)

Directory filter is a regular select html element:
DeepinScreenshot_select-area_20200131162453

**TODO:** 
1. Add styles to match mockup **done** :heavy_check_mark: 
2. Resolve "children" directories to be displayed correctly : **done** :heavy_check_mark: 
3. Setup filter on "click" event **done** :heavy_check_mark:

### CODE REVIEW

File: MediaGalleryUi/view/adminhtml/templates/directory_container.phtml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    {
+        ".media-directory-container": {
+            "Magento_MediaGalleryUi/js/directory/directoryTree": {
+                "directoryListUrl": "<?= $block->getUrl("media_gallery/directories/getlist"); ?>"
+            }
+        }
+    }
+</script>
+<!-- @todo remove when "UI components" will support such initialization -->
+<script>
```
Comment (sivaschenko) [inline MediaGalleryUi/view/adminhtml/templates/directory_container.phtml]: I think its better to use jquery widget as proxy to UI components like we do with panels and grids: i.e. `MediaGalleryUi/view/adminhtml/web/js/container.js`

File: MediaGalleryUi/Controller/Adminhtml/Directories/GetTree.php [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        $this->responseContent = [];
+    }
+    /**
+     * @inheritdoc
+     */
+    public function execute()
+    {
+        try {
+            $directories = [];
+            $directoryInstance = $this->filesystem->getDirectoryRead($this->path);
```
Comment (sivaschenko) [inline MediaGalleryUi/Controller/Adminhtml/Directories/GetTree.php]: Please extract the logic to a model

File: MediaGalleryUi/view/adminhtml/ui_component/media_gallery_listing.xml [hunk 3]
```diff
@@ -112,5 +114,5 @@
                 <sortable>false</sortable>
             </settings>
         </column>
-    </columns>
+    </columns>
```
Comment (sivaschenko) [inline MediaGalleryUi/view/adminhtml/ui_component/media_gallery_listing.xml]: Please remove extra spaces

File: MediaGalleryUi/view/adminhtml/web/js/grid/columns/image.js [hunk 4]
```diff
@@ -11,6 +11,7 @@ define([
         defaults: {
             bodyTmpl: 'Magento_MediaGalleryUi/grid/columns/image',
             selected: null,
+            getDirectoryTreeUrl: 'media_gallery/directories/gettree',
```
Comment (sivaschenko) [inline MediaGalleryUi/view/adminhtml/web/js/grid/columns/image.js]: The image component is used to render image column, it should not be aware of directories tree

**Review by diazwatson — DISMISSED** [review]

**Comment by diazwatson** [comment]
Forgot to add my screen:
Screenshot 2020-02-06 at 16 37 28

**Comment by Nazar65** [comment]
@diazwatson wow, looks like styles is missing, will recheck

**Comment by diazwatson** [comment]
Thanks @Nazar65 let me know when you ready for me to re test

**Comment by Nazar65** [comment]
@diazwatson well after you switched to this branch you need to re deploy static content, and all related styles should be added
DeepinScreenshot_select-area_20200212155712

File: MediaGalleryUi/Ui/Component/Listing/Columns/Url.php [hunk 5]
```diff
@@ -24,23 +24,49 @@ class Url extends Column
     private $storeManager;
 
     /**
+     * @var UrlInterface
+     */
+    private $url;
```
Comment (sivaschenko) [inline MediaGalleryUi/Ui/Component/Listing/Columns/Url.php]: This column should not be aware of directory tree. A dedicated `class` can be assigned to the directory tree container in the xml to prepare the URL

File: MediaGalleryUi/view/adminhtml/web/css/source/_module.less [hunk 6]
```diff
@@ -6,6 +6,44 @@
...
+            background-position: -52px -16px;
+        }
+
+        .jstree a > ins {
+            height: 20px;
+            width: 20px;
+        }
+
+        .jstree-default .jstree-no-dots .jstree-leaf > ins{ 
+            background-image: none !important;
```
Comment (sivaschenko) [inline MediaGalleryUi/view/adminhtml/web/css/source/_module.less]: please avoid `!important` in the css

File: MediaGalleryUi/view/adminhtml/web/css/source/_module.less [hunk 7]
```diff
@@ -6,6 +6,44 @@
...
+            height: 20px;
+            width: 20px;
+        }
+
+        .jstree-default .jstree-no-dots .jstree-leaf > ins{ 
+            background-image: none !important; 
+        }
+
+        .jstree-default ins {
+            background-image: url("../../images/d.png") !important;
```
Comment (sivaschenko) [inline MediaGalleryUi/view/adminhtml/web/css/source/_module.less]:
```suggestion
            background-image: url("@{baseDir}Magento_MediaGalleryUi/images/d.png");
```

File: MediaGalleryUi/Controller/Adminhtml/Directories/GetTree.php [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+    /**
+     * @var string
+     */
+    private $path;
+
+    /**
+     * @var array
+     */
+    private $responseContent;
```
Comment (sivaschenko) [inline MediaGalleryUi/Controller/Adminhtml/Directories/GetTree.php]: This property is not necessary, a local variable should be used instead

File: MediaGalleryUi/Controller/Adminhtml/Directories/GetTree.php [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+        return $resultJson;
+    }
+
+    /**
+     * Build folder tree structure by provided directories path
+     *
+     * @param array $directories
+     * @return array
+     */
+    private function buildFolderTree(array $directories, bool $skipRoot = true): array
```
Comment (sivaschenko) [inline MediaGalleryUi/Controller/Adminhtml/Directories/GetTree.php]: Is it possible to perform the final folder tree creation in the directory iterator loop? (for better performance)

File: MediaGalleryUi/Controller/Adminhtml/Directories/GetTree.php [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+
+    /**
+     * Find parent directory
+     *
+     * @param array $node
+     * @param array $treeNode
+     * @param int $level
+     * @return array
+     */
+    private function findParent(array &$node, array &$treeNode, int $level = 0): array
```
Comment (sivaschenko) [inline MediaGalleryUi/Controller/Adminhtml/Directories/GetTree.php]: I believe the custom implementation of recursive directory iteration would help to reduce the algorithm complexity (no need to search for parent if the algorithm builds tree configuration level by level).

File: MediaGalleryUi/view/adminhtml/web/js/directory/directoryTree.js [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    'use strict';
+
+    return Component.extend({
+        defaults: {
+            filterChipsProvider: 'componentType = filters, ns = ${ $.ns }',
+            directoryTreeSelector: '#media-gallery-directory-tree',
+            /* eslint-disable max-len */
+            urlProvider: 'name = media_gallery_listing.media_gallery_listing.media_gallery_columns.thumbnail_url, ns = media_gallery_listing',
+            options: {
+                treeInitData: []
```
Comment (sivaschenko) [inline MediaGalleryUi/view/adminhtml/web/js/directory/directoryTree.js]: Why do we need `options` parent node? Are there any other options that we might use?

File: MediaGalleryUi/view/adminhtml/web/js/directory/directoryTree.js [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+
+        /**
+         * Get json data for jstree
+         */
+        getJsonTree: function () {
+            $.ajax({
+                url: this.image().getDirectoryTreeUrl,
+                type: 'POST',
+                dataType: 'json',
+                async: false,
```
Comment (sivaschenko) [inline MediaGalleryUi/view/adminhtml/web/js/directory/directoryTree.js]: I think this ajax call can be done asynchronously if `createTree` is called in the success callback

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Thanks for the pull request @Nazar65 ! Please see my review comments

File: MediaGalleryUi/Controller/Adminhtml/Directories/GetTree.php (hunk 2 above)
Comment (Nazar65) [reply MediaGalleryUi/Controller/Adminhtml/Directories/GetTree.php]: ok

File: MediaGalleryUi/view/adminhtml/ui_component/media_gallery_listing.xml (hunk 3 above)
Comment (Nazar65) [reply MediaGalleryUi/view/adminhtml/ui_component/media_gallery_listing.xml]: ok

**Pull request merged**